_posts/2025-03-04-fast-streaming-inserts-in-duckdb-with-adbc.md

```diff
@@ -0,0 +1,201 @@
+
+# Results
+
+<img src="{{ site.baseurl }}/img/adbc-duckdb/btop.png" width="100%" class="img-responsive" alt="btop utility showing CPU and memory usage of quacfka-service and runner" aria-hidden="true">  
+Note: both runs with GOMAXPROCS set to 24 (the number of DuckDB insertion routines)
+
+Ingesting the raw data (14 fields with one deeply nested LIST.STRUCT.LIST field) \+ normalized data:  
+  *num\_cpu: 60*  
+  *runtime\_os: linux*  
+  *kafka\_clients: 5*  
+  *kafka\_queue\_cap: 983040*  
+  *processor\_routines: 32*  
+  *arrow\_queue\_cap: 4*  
+  *duckdb\_threshold\_mb: 4200*  
+  *duckdb\_connections: 24*  
+  *normalizer\_fields: 10*  
+  *start\_time: 2025-02-24T21:06:23Z*  
+  *end\_time: 2025-02-24T21:11:23Z*  
+  *records: 123\_686\_901.00*  
+  *norm\_records: 122\_212\_452.00*  
+  *data\_transferred: 146.53 GB*  
+  *duration: 4m59.585s*  
+  ***records\_per\_second: 398\_271.90***  
+  ***total\_rows\_per\_second: 806\_210.41***  
+  *transfer\_rate: 500.86 MB/second*  
+  *duckdb\_files: 9*  
+  *duckdb\_files\_MB: 38429*  
+  *file\_avg\_duration: 33.579s*
+
+How many rows/second could we get if we only inserted the flat, normalized data? (Note: original records are still processed, just not inserted) 
+  *num\_cpu: 60*  
+  *runtime\_os: linux*  
+  *kafka\_clients: 10*  
+  *kafka\_queue\_cap: 1228800*  
+  *processor\_routines: 32*  
+  *arrow\_queue\_cap: 4*  
+  *duckdb\_threshold\_mb: 4200*  
+  *duckdb\_connections: 24*  
+  *normalizer\_fields: 10*  
+  *start\_time: 2025-02-25T19:04:33Z*  
+  *end\_time: 2025-02-25T19:09:36Z*  
+  *records: 231\_852\_772.00*  
+  *norm\_records: 363\_247\_327.00*  
+  *data\_transferred: 285.76 GB*  
+  *duration: 5m3.059s*  
+  *records\_per\_second: 0.00*  
+  ***total\_rows\_per\_second: 1\_198\_601.39***  
+  *transfer\_rate: 965.54 MB/second*  
+  *duckdb\_files: 5*  
+  *duckdb\_files\_MB: 20056*  
+  *file\_avg\_duration: 58.975s*  
+<img src="{{ site.baseurl }}/img/adbc-duckdb/onemillionrows.png" width="100%" class="img-responsive" alt="One million rows/second" aria-hidden="true"> 
+
+Once deployed, the number of parquet files fall from ~3000 small files per hour to ~12 files per hour. Goodbye small files!
+<img src="{{ site.baseurl }}/img/adbc-duckdb/kip_yes.gif" width="25%" class="img-responsive" alt="Yesss" aria-hidden="true"> 
+
+# Challenges/Learnings
```

Comment:
Can you please end each of these bullet points with a period, since they are all complete sentences? Thanks